ocaml/testsuite/tests/typing-layouts/basics_beta.ml

```diff
@@ -545,3 +545,26 @@ Line 3, characters 15-40:
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
 Error: Type 'a has layout value, which is not a sublayout of immediate.
 |}]
+
+(****************************************************)
+(* Test 35: check bad layout error in filter_arrow *)
+
+module M35_1 : sig
+  val x : string
+end = struct
+  type ('a : immediate) t = 'a
+
+  let f : 'a t -> 'a = fun x y z -> x
+
+  let x = f (assert false)
+
+end;;
+
+[%%expect {|
+Line 6, characters 23-37:
+6 |   let f : 'a t -> 'a = fun x y z -> x
+                           ^^^^^^^^^^^^^^
+Error: This expression has type 'a -> 'b
```

Comment:
This error message is pretty bad. I suppose it's better than tthrowing an exception. But can we do better before all the layout histories work is merged?

Reply:
Does dropping the `Diff` element of the trace list make the error any better?  If not, one could consider adding a new `filter_arrow_failure` case.  The revised one may be good enough for now, though.

Reply:
Dropping the `Diff` element results in a similar error but with the first two lines removed:

```
> --- basics_beta.ml	2023-09-11 16:14:09.400891238 -0400
> +++ basics_beta.ml.corrected.corrected	2023-09-11 16:44:50.991568189 -0400
> @@ -557,7 +557,6 @@
>  Line 2, characters 19-31:
>  2 | let f35 : 'a t35 = fun () -> ()
>                         ^^^^^^^^^^^^
> -Error: This expression has type 'b -> 'c
> -       but an expression was expected of type ('a : immediate)
> +Error:
>         'a -> 'b has layout value, which is not a sublayout of immediate.
>  |}]
> 
> )
```

Reply:
This looks like an improvement to me, since the problem with the old error was, in part, that it was talking about a difference between two types that didn't obviously correspond to something meaningful.  But I wonder what Richard thinks.

Reply:
Yes, I think this smaller error is better. Though there is a weird extra newline after the `Error:` that might be worth figuring out. (Sometimes tracking down these weird inconsistencies reveals something interesting. Sometimes it's just an extra `\n` though.)